File: persistence/src/main/java/de/cyface/persistence/model/DataPoint.java
package de.cyface.persistence.model;

import android.os.Parcel;
import android.os.Parcelable;

/**
 * An abstract base class for all data points processed by Cyface. It provides the generic functionality of a data point
 * to be unique on this device and to have a Unix timestamp associated with it.
 *
 * @author Klemens Muthmann
 * @version 1.0.0
 * @since 1.0.0
 */
public abstract class DataPoint implements Parcelable {
    /**
     * The system wide unique identifier for this {@code DataPoint}. If the point was not saved yet it may be
     * {@code null}, since the persistence layers assigns a unique identifier on saving a data point.
     * For AndroidDataAccessLayer.getIdOfNextUnSyncedMeasurement() to work the id must be long ASC
     */
    private Long identifier;
    /**
     * The Unix timestamp at which this {@code DataPoint} was measured in milliseconds.
     */
    private final long timestamp;

    /**
     * Creates a new completely initialized {@code DataPoint}.
     *
     * @param identifier The system wide unique identifier for this {@code DataPoint}.
     * @param timestamp The Unix timestamp at which this {@code DataPoint} was measured in milliseconds.
     */
    public DataPoint(final Long identifier, final long timestamp) {
        if(timestamp<0L) {
            throw new IllegalArgumentException("Illegal argument: timestamp was less than 0L!");
        }

        this.identifier = identifier;
        this.timestamp = timestamp;
    }

    /**
     * @return The system wide unique identifier for this {@code DataPoint}. If the point was not saved yet it may be
     *         {@code null}, since the persistence layers assigns a unique identifier on saving a data point.
     */
    public long getIdentifier() {
        return identifier;
    }

    /**
     * @return The Unix timestamp at which this {@code DataPoint} was measured.
     */
    public long getTimestamp() {
        return timestamp;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }

        DataPoint dataPoint = (DataPoint)o;

        return identifier == dataPoint.identifier;

    }

    @Override
    public int hashCode() {
        return (int)(identifier ^ (identifier >>> 32));
    }

    /*
     * MARK: Code for Parcelable interface.
     */

    /**
     * Recreates this point from the provided <code>Parcel</code>.
     *
     * @param in Serialized form of a <code>DataPoint</code>.
     */
    public DataPoint(Parcel in) {
        this.identifier = (Long)in.readValue(getClass().getClassLoader());
        this.timestamp = in.readLong();
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeValue(identifier);
        dest.writeLong(timestamp);
    }
}

File: persistence/src/main/java/de/cyface/persistence/model/DataPoint.kt
/*
 * Copyright 2023 Cyface GmbH
 *
 * This file is part of the Cyface SDK for Android.
 *
 * The Cyface SDK for Android is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * The Cyface SDK for Android is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with the Cyface SDK for Android. If not, see <http://www.gnu.org/licenses/>.
 */
package de.cyface.persistence.model

import android.os.Parcel
import android.os.Parcelable
import java.util.Objects

/**
 * An abstract base class for all data points processed by Cyface. It provides the generic functionality of a data point
 * to be unique on this device and to have a Unix timestamp associated with it.
 *
 * @author Armin Schnabel
 * @version 2.0.0
 * @since 1.0.0
 * @property timestamp The Unix timestamp at which this [DataPoint] was measured in milliseconds.
 */
abstract class DataPoint(open val timestamp: Long) : Parcelable {
    init {
        require(timestamp >= 0L) { "Illegal argument: timestamp was less than 0L!" }
    }

    /*
     * MARK: Code for Parcelable interface.
     */

    /**
     * Recreates this point from the provided `Parcel`.
     *
     * @param in Serialized form of a [DataPoint].
     */
    protected constructor(`in`: Parcel) : this(`in`.readLong())

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeLong(timestamp)
    }

    override fun toString(): String {
        return "DataPoint{" +
                "timestamp=" + timestamp +
                '}'
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val that = other as DataPoint
        return timestamp == that.timestamp
    }

    override fun hashCode(): Int {
        return Objects.hash(timestamp)
    }
}